zipline/data/_equities.pyx

```diff
@@ -170,22 +176,42 @@ cpdef _read_bcolz_data(ctable_t table,
         intp_t first_row
         intp_t last_row
         intp_t offset
+        intp_t out_start
+        intp_t out_end
         list results = []
 
+    ndays = shape[0]
     nassets = shape[1]
     if not nassets== len(first_rows) == len(last_rows) == len(offsets):
         raise ValueError("Incompatible index arrays.")
 
     for column_name in columns:
-        raw_data = table[column_name][:]
         outbuf = zeros(shape=shape, dtype=uint32)
-        for asset in range(nassets):
-            first_row = first_rows[asset]
-            last_row = last_rows[asset]
-            offset = offsets[asset]
-            for out_idx, raw_idx in enumerate(range(first_row, last_row + 1)):
-                outbuf[out_idx + offset, asset] = raw_data[raw_idx]
-
+        if read_all:
+            raw_data = table[column_name][:]
+
+            for asset in range(nassets):
+                first_row = first_rows[asset]
+                last_row = last_rows[asset]
+                offset = offsets[asset]
+                for out_idx, raw_idx in enumerate(
+                        range(first_row, last_row + 1)):
+                    outbuf[out_idx + offset, asset] = raw_data[raw_idx]
+        else:
+            carray = table[column_name]
+
+            for asset in range(nassets):
+                first_row = first_rows[asset]
+                last_row = last_rows[asset]
+                offset = offsets[asset]
+                out_start = offset
+                out_end = (last_row - first_row) + offset + 1
```

Comment:
The following has similar performance of around 100ms when forced into `read_all` mode (tested with assets length of 1, 1000, 5000, 7000). The equivalency makes sense, since the write to a contiguous slice should be performant.

```
diff --git a/zipline/data/_equities.pyx b/zipline/data/_equities.pyx
index ef6bba6..caa8ae6 100644
--- a/zipline/data/_equities.pyx
+++ b/zipline/data/_equities.pyx
@@ -193,8 +193,11 @@ cpdef _read_bcolz_data(ctable_t table,
                 first_row = first_rows[asset]
                 last_row = last_rows[asset]
                 offset = offsets[asset]
-                for out_idx, raw_idx in enumerate(range(first_row, last_row + 1)):
-                    outbuf[out_idx + offset, asset] = raw_data[raw_idx]
+                if first_row <= last_row:
+                    outbuf[offset:offset + (last_row + 1 - first_row), asset] =\
+                        raw_data[first_row:last_row + 1]
+                else:
+                    continue
         else:
             carray = table[column_name]
```

@ssanderson shall we go with applying this patch to make the implementation more consistent?